core_lib/src/structure/copymultiplekeyframesdialog.cpp

```diff
@@ -0,0 +1,403 @@
+#include "copymultiplekeyframesdialog.h"
+#include "ui_copymultiplekeyframesdialog.h"
+#include "timeline.h"
+#include "timecontrols.h"
+#include "layer.h"
+#include "layermanager.h"
+#include <QDebug>
+#include <QSettings>
+
+CopyMultiplekeyframesDialog::CopyMultiplekeyframesDialog(QWidget *parent) :
+    QDialog(parent),
+    ui(new Ui::CopyMultiplekeyframesDialog)
+{
+    ui->setupUi(this);
+    init();
+}
+
+
+CopyMultiplekeyframesDialog::CopyMultiplekeyframesDialog(LayerManager *lm, int startLoop, int stopLoop, QWidget *parent):
+    QDialog(parent),
+    ui(new Ui::CopyMultiplekeyframesDialog)
+{
+    mFirstFrame = startLoop;
+    mLastFrame = stopLoop;
+    lMgr = lm;
+    ui->setupUi(this);
+    setWindowTitle(tr("Manipulate Range of Frames"));
+    init();
+}
+
+CopyMultiplekeyframesDialog::~CopyMultiplekeyframesDialog()
+{
+    delete ui;
+}
+
+void CopyMultiplekeyframesDialog::init()
+{
+    // SET values in spinBoxes
+    ui->sBoxFirstFrame->setValue(mFirstFrame);
+    ui->sBoxLastFrame->setValue(mLastFrame);
+    ui->sBoxStartFrame->setValue(mLastFrame + 1);
+    ui->sBoxMove->setValue(mLastFrame + 1);
+    ui->sBoxStartReverse->setValue(mLastFrame + 1);
+
+    // SET values in ComboBoxes
+    int lType = lMgr->currentLayer()->type(); // find layer-type
+    for (int i = 1; i < lMgr->count(); i++)
+    {
+        if (lMgr->getLayer(i)->type() == lType)
+        {
+            ui->cBoxFromLayer->addItem(lMgr->getLayer(i)->name());
+            ui->cBoxCopyToLayer->addItem(lMgr->getLayer(i)->name());
+            ui->cBoxMoveToLayer->addItem(lMgr->getLayer(i)->name());
+        }
+    }
```

Comment:
As mentioned above, I have never worked professionally as a programmer. On the other hand I have 20+ years behind me as a professional animator, working on features, TV-series, commercials and so on. 
My urge to help coding, is that I want to use Pencil2D professionally, and if my coding skills can help Pencil2D to move up where it can (and will) be used by amateurs and professionals alike, then I'm ready to spend the time. 
When I'm coding, I think very much in terms of an animators work process. In this example, If I was working on a Bitmap layer, I would not get an urge to manipulate frames on a Vector layer. In any case, that's not the way I work.

Reply:
I get your line of thought. My line of thought was that we also have a lot of not-so-technical users to whom the difference between vector and bitmap isn’t that clear-cut, at least I seem to remember hearing questions about that somewhere. But maybe @Jose-Moreno can chime in here, I think he has a better idea of what our userbase is like than just about anyone else.

Reply:
@J5lx @davidlamhauge Sorry didn't receive a notification, and just saw this msg. 

In this case I agree with David's concept. as far as I saw in his demo, the layer type is automatically filtered to only show the relevant layer type you're working on. To me that is more than ideal, because even if the difference between both layer types isn't as clear cut as you say, the problem here isn't understanding but functionality, and I know people will mess up by selecting the wrong layer type time and time again if we allow them to see all of them, which would also impact their performance on a job.

This would perhaps be a real consideration if Pencil2D had some sort of raster to vector to raster conversion, and even then I feel that such conversion should happen outside of the scope of this function, so it's better to have the displayed layers filtered like they are now, to avoid unnecessary confusion and improve workflow efficiency.

----
**Note:** I think It's important that we establish that the difference between layer types isn't as pronounced mainly because in a production both can be used separately or in combination to aid the production process.  I've mentioned this before, but I'll lay it out here so I can come back to it later.

The premise is that the bitmap layer is for raster animation and allows you to go through a full production pipeline from storyboarding to final animation (except compositing and editing) and it will allow several art styles but more commonly a "painterly" style if the user chooses to finish the work using only a bitmap layer (software with similar focus: TVPaint, Krita, Photoshop)

The vector layer however can be used for the same pipeline as raster animation, but the applicable styles are limited on inking and coloring style. Normally the "flat-style" seen in TV productions which use mono-width line strokes and cel-shading. Currently Pencil2D accepts variable line width for vector strokes and it will be proposed we consider having gradient vector fills to improve the range of artistic styles for this layer type. This workflow strength is the ink (outline drawing) & paint (fill coloring) stages and capitalizes on the efficiency of application and modification due to the vector nature. (software with similar focus: Flash, Toonboom, Opentoonz) There is also a discussion pending on interpolated inbetweening for this layer type.

Lastly both layer types can be used together to highlight the strengths of both workflows using raster for pre-production planning and rough animation, while using vector for the clean-up and coloring stages of a pipeline allowing the computer to assist the animators to improve their efficiency during different stages of work.

Reply:
To clarify a little, what I originally intended was that only the source combobox (“from”) allows any layer type to be selected, while the target combobox (“to”) still *shows* all layers but allows only those of the same type as the “from” layer to be selected, while the others are disabled. That way, what the user can do with the dialog is not impacted by the layer selected when the dialog was opened. The way it is now, the selected layer does influence what the user can do with the dialog, but only indirectly (the user can still alter the source layer, but to select a source layer of a type other than the one of the initial source layer the dialog has to be re-opened), which I thought might be confusing. That said, you made a good point regarding efficiency and reduced complexity. Thanks for the elaborate response!